src/openapi/streaming/connection/transport/websocket-transport.js

```diff
@@ -181,58 +189,34 @@ function handleSocketMessage(messageEvent) {
 }
 
 function handleSocketClose(event) {
-    if (
-        (this.socket && this.socket.readyState === STATE_INTERNAL_CLOSED) ||
-        this.socket.readyState === STATE_INTERNAL_CLOSING
-    ) {
-        log.debug(LOG_AREA, 'Transport connection closed', {
-            readyState: this.socket.readyState,
-            code: event.code,
-        });
-
-        // 1006 doesn't tell us its a 401 unauthorized, just that it might be
-        // the websocket spec disallows reading status codes for security reasons
-        // (so websocket can't be used to probe http endpoints)
-        // So, before we reconnect, lets re-authorize to be sure
-        // and if that gets a 401, that will report the token invalid via
-        // authTransport
-        if (event.code === 1006) {
-            this.getAuthorizePromise(this.contextId, true);
-        }
-
-        // Order here is important. We need to invoke state callback first, as parent streaming manager updates query upon disconnect.
-        this.stateChangedCallback(constants.CONNECTION_STATE_DISCONNECTED);
-        reconnect.call(this);
+    if (!this.socket) {
+        return;
     }
-}
 
-function handleSocketOpen() {
-    if (this.socket && this.socket.readyState === STATE_INTERNAL_OPEN) {
-        this.hasWorked = true;
-        this.reconnectCount = 0;
+    if (!this.hasWorked) {
+        handleFailure.call(this, {
+            message: `websocket error occured. code: ${event.code}, reason: ${event.reason}`,
+        });
 
-        log.debug(LOG_AREA, 'Socket opened');
-        this.stateChangedCallback(constants.CONNECTION_STATE_CONNECTED);
+        return;
     }
-}
 
-function handleSocketError(error) {
-    if (this.socket) {
-        log.error(LOG_AREA, 'Socket error occurred', {
-            code: error.code,
-            reason: error.reason,
+    const isCleanDisconnect = event.wasClean === true;
+    if (!isCleanDisconnect) {
+        log.error(LOG_AREA, 'websocket connection closed abruptly', {
+            readyState: this.socket.readyState,
+            code: event.code,
+            reason: event.reason,
         });
+    } else {
+        log.debug(LOG_AREA, 'websocket connection closed');
+    }
 
-        if (this.hasWorked) {
-            // If transport worked at least once, try to reconnect when error occurs instead of failing.
-            this.stateChangedCallback(constants.CONNECTION_STATE_DISCONNECTED);
-            reconnect.call(this);
-        } else {
-            handleFailure.call(this, {
-                message: `WebSocket connection failed with error code: ${error.code}, reason: ${error.reason}`,
-            });
-        }
+    if (event.code === socketCloseCodes.TOKEN_EXPIRED) {
+        this.unauthorizedCallback();
     }
+
+    reconnect.call(this);

```

Comment:
Reconnecting immediately after calling this.unauthorizedCallback doesn't look like a good approach. If current token is invalid, new connect will still fail and might cause spam of oauth token re-requests.

I think instead we should postpone reconnect here until fresh token is received.

Reply:
reconnection happens after reconnect delay
we were relying on timer for authorisation before this change as well
I have already discussed this on slack with you

Reply:
I don't think previous approach was right :) Can we try to fix it by making reconnect logic wait for authorization response? When we receive new token, we currently call: WebsocketTransport.prototype.updateQuery.
This would be indicator that we received new token after token experied close code and we are ready to continue re-connection. 

WDYT?

Reply:
Done